docs/versioned_docs/version-zenith/quickstart/292472-inputs-outputs.mdx

```diff
@@ -0,0 +1,53 @@
+---
+slug: /quickstart/292472/inputs-outputs
+hide_table_of_contents: true
+title: "Understand function inputs and outputs"
+---
+import PartialExperimentalDocs from '../partials/_experimental.mdx';
+
+# Quickstart
+
+<PartialExperimentalDocs />
+
+## Understand function inputs and outputs
+
+
+Module functions can accept arguments and return both basic and custom types. Dagger introduces valuable new primitives such as `Directory`, `Container`, and `Service`, and Dagger functions can use these as both inputs and outputs (in addition to your own custom types).
+
+To begin understanding this, call a function using the Dagger CLI:
+
+```shell
+dagger call -m github.com/kpenfound/dagger-modules/golang --proj https://github.com/dagger/dagger test --args ./core
+```
+
+This Go builder module exposes a `Test()` function that accepts a remote Go repository URL as input argument. It then runs the test suite in the repository and returns the test results to the caller. This example uses Dagger's GitHub repository as input and returns the results of a subset of Dagger's own test suite.
+
+Once the command completes, you should see output similar to the following:
+
+```shell
+✔ dagger call test [5m19.0s]
+┃ ok      github.com/dagger/dagger/core   0.397s
+```
+
+Dagger functions can also return directories. To see this in action, use the following command:
+
+```shell
+dagger call -m github.com/kpenfound/dagger-modules/golang --proj https://github.com/dagger/dagger build --args ./cmd/dagger
+```
+
+This example uses the same Go builder module as before, but this time calls its `Build()` function. Like the `Test()` function, the `Build()` function accepts a remote Go repository URL and builds a binary from the remote Go repository. It returns a Dagger `Directory` type representing the state of the working directory in the repository after the build, similar to the output shown below:
```

Comment:
On one hand, I had planned to restrict this page to calls returning primitives, and moving all discussion of chaining to the next page. OTOH, for the container example on this page, I had to chain calls anyway, because I couldn't find a module that returns a container directly without needing a chained call - is there one? Still trying to find a nice way to reconcile this discrepancy

Reply:
Yeah, in the python dev module.

`all-tests` return a list of two test instances (for Python 3.10 and Python 3.11 versions):
https://github.com/dagger/dagger/blob/6a3634f6e9b66c124511e07ffc669e3fe5e709ed/sdk/python/dev/src/main/__init__.py#L100

Each test has `with_focus` before the `with_exec`, so it'll stream the output to the TUI as it's happening, and at the same time:

https://github.com/dagger/dagger/blob/6a3634f6e9b66c124511e07ffc669e3fe5e709ed/sdk/python/dev/src/main/test.py#L58-L60

The list and with_focus is not important. Most of the time, with tests, you just want to know if it succeeded or not. If it failed you want the error. So a simple: `dagger call test`, where _test_ is a container that has a `with_exec` that executes the test suite, it'll sync and run and just show that it's in progress. But if it fails, it'll present the error since that's how the TUI behaves on errors, even when `--focus=true`.

Reply:
Note, I address this discrepancy in my proposed edits.